### PR TITLE
Make balpy feature complete with the Vault

### LINKED ISSUE #3: Add pool exits
Currently the library does not support pool exits, which would be a nice addition! The code for joins should be relatively similar, so looking at those could be a good starting point

### PR DESCRIPTION
**Pull request by ribeirojose**
closes #3 and #5

### CODE REVIEW

File: balpy/balpy.py [hunk 1]
```diff
@@ -1510,11 +1515,195 @@ def balLinearPoolInitJoin(self, poolDescription, poolId, slippageTolerancePercen
...
+        ):
+		tokenList, minAmountsOut = self.balSortTokensExitPoolExactBptInForTokensOut(
+		    tokenList, minAmountsOut)
+		userData = eth_abi.encode_abi(["uint256", "uint256"], [
+		                              exitKindValue, bptAmount])
+		return tokenList, minAmountsOut, userData, toInternalBalance
+
+	def balSortTokensExitPoolBptInForExactTokensOut(self, tokenList, amountsOut, minAmountsOut):
+		tokenListAddresses = [self.web3.toChecksumAddress(
+		    token) for token in tokenList]
```
Comment (gerrrg) [inline balpy/balpy.py:1615]:
```suggestion
		tokenListAddresses = [self.web3.toChecksumAddress(token) for token in tokenList]
```

File: balpy/balpy.py [hunk 2]
```diff
@@ -1510,11 +1515,195 @@ def balLinearPoolInitJoin(self, poolDescription, poolId, slippageTolerancePercen
...
+		minAmountsOutSorted = [
+			minAmountOut
+			for _, minAmountOut in sorted(zip(tokenListAddresses, minAmountsOut))
+		]
+		amountsOutSorted = [
+			amountOut for _, amountOut in sorted(zip(tokenListAddresses, amountsOut))
+		]
```
Comment (gerrrg) [inline balpy/balpy.py:1622]: make more Pythonic

File: balpy/balpy.py [hunk 3]
```diff
@@ -1510,11 +1515,195 @@ def balLinearPoolInitJoin(self, poolDescription, poolId, slippageTolerancePercen
...
+		return tokenListAddressSorted, amountsOutSorted, minAmountsOutSorted
+
+	def balFormatExitPoolRequestTupleBptInForExactTokensOut(
+                self,
+                exitKindValue,
+                tokenList,
+                maxBptAmount,
+                amountsOut,
+                minAmountOut,
+                toInternalBalance=True):
```
Comment (gerrrg) [inline balpy/balpy.py:1633]: why is `toInternalBalance` defaulting to `True`?

File: balpy/balpy.py [hunk 4]
```diff
@@ -1510,11 +1515,195 @@ def balLinearPoolInitJoin(self, poolDescription, poolId, slippageTolerancePercen
...
 	def balVaultWeth(self):
 		vault = self.balLoadContract("Vault");
 		wethAddress = vault.functions.WETH().call();
 		return(wethAddress);
 
+	def balVaultGetAuthorizer(self):
+		vault = self.balLoadContract("Vault")
+		authorizerAddress = vault.functions.getAuthorizer().call()
+		return authorizerAddress
+
```
Comment (gerrrg) [inline balpy/balpy.py:1706]: write quick sample in samples/misc/

File: balpy/balpy.py [hunk 5]
```diff
@@ -1510,11 +1515,195 @@ def balLinearPoolInitJoin(self, poolDescription, poolId, slippageTolerancePercen
...
+            self,
+            poolId: str,
+            bptAmount: float,
+            tokenList: List[str],
+            exitKind: str,
+            tokenOut: int = None,
+            amountsOut: List[int] = [],
+            query: bool = False,
+            toInternalBalance: bool = True,
+            minAmountsOut: List[float] = None,
```
Comment (gerrrg) [inline balpy/balpy.py:1651]: please follow convention laid out by `balJoinPool` or update the rest of the repo to follow this `**exitData` pattern

File: samples/exitPool/sampleExitBptInForExactTokensOut.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+  "network": "goerli",
+  "poolId": "0x9f1f16b025f703ee985b58ced48daf93dad2f7ef000200000000000000000063",
+  "bptAmount": 0.5,
```
Comment (gerrrg) [inline samples/exitPool/sampleExitBptInForExactTokensOut.json:4]: add `toInternalBalance` as an argument here, defaulted to `false`

File: samples/exitPool/sampleExitExactBptInForOneTokenOut.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "network": "goerli",
+  "poolId": "0x9f1f16b025f703ee985b58ced48daf93dad2f7ef000200000000000000000063",
+  "bptAmount": 0.1,
```
Comment (gerrrg) [inline samples/exitPool/sampleExitExactBptInForOneTokenOut.json:4]: add `toInternalBalance` as an argument here, defaulted to `false`

File: samples/exitPool/sampleExitExactBptInForTokensOut.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+  "network": "goerli",
+  "poolId": "0x9f1f16b025f703ee985b58ced48daf93dad2f7ef000200000000000000000063",
+  "bptAmount": 0.1,
+  "tokenList": [
```
Comment (gerrrg) [inline samples/exitPool/sampleExitExactBptInForTokensOut.json:5]: rename list to `tokens`

File: samples/exitPool/sampleExitExactBptInForOneTokenOut.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  "tokenList": [
```
Comment (gerrrg) [inline samples/exitPool/sampleExitExactBptInForOneTokenOut.json:5]: rename list to `tokens`

**Review by gerrrg — CHANGES_REQUESTED** [review]
Overall looks pretty good. I tested the exit samples successfully (though I was confused at first since everything was going to internal balance). Most comments revolve around:
* maintaining patterns of the repo
* keeping code Pythonic
* writing samples for some of the easy functions

**Comment by ribeirojose** [comment]
Hey @gerrrg just letting you know we've amended the changes and this is ready for a second look! Let me know if you want us to resolve the conversations in the PR or if you'd prefer to do it yourself 😄

**Comment by gerrrg** [comment]
@ribeirojose nice! Just responding now to acknowledge that I've seen this, and I'll try to take a look through the changes soon to get this back to you quickly

File: balpy/balpy.py [hunk 10]
```diff
@@ -98,7 +98,7 @@ class balpy(object):
 						"ropsten":	{"id":3,		"blockExplorerUrl":"ropsten.etherscan.io"													},
 						"rinkeby":	{"id":4,		"blockExplorerUrl":"rinkeby.etherscan.io"													},
 						"goerli":	{"id":5,		"blockExplorerUrl":"goerli.etherscan.io"													},
-						"optimism":	{"id":10,		"blockExplorerUrl":"optimistic.etherscan.io"													},
+						"optimism":	{"id":10,		"blockExplorerUrl":"optimistic.etherscan.io"												},
```
Comment (gerrrg) [inline balpy/balpy.py:101]: Is this just `diff` being weird, or did something actually change here?

File: balpy/balpy.py [hunk 11]
```diff
@@ -1766,11 +1687,11 @@ def balVaultSetPaused(self, paused, isAsync, **buildTxKwargs):
 		tx = self.buildTx(setPausedFn, **buildTxKwargs)
 		return self.sendTx(tx, isAsync)
 
-	def balVaultSetRelayerApproval(self, senderAddress, relayerAddress, approved, isAsync=False, **buildTxKwargs):
+	def balVaultSetRelayerApproval(self, senderAddress, relayerAddress, approved, isAsync=False, gasFactor=1.05, **buildTxKwargs):
```
Comment (gerrrg) [inline balpy/balpy.py:1690]: one lingering `**<kwargs>` here

File: samples/exitPool/sampleExitBptInForExactTokensOut.json [hunk 12]
```diff
@@ -1,11 +1,17 @@
...
-  ],
-  "amountsOut": [0.001, 1],
-  "exitKind": "BPT_IN_FOR_EXACT_TOKENS_OUT"
+  "bptAmount": "0.5",
+  "exitKind": "BPT_IN_FOR_EXACT_TOKENS_OUT",
+  "toInternalBalance": false,
+  "tokens": {
+    "0xdfcea9088c8a88a76ff74892c1457c17dfeef9c1": {
+      "amount": "0.001",
+      "minAmount": "0.0009"
```
Comment (gerrrg) [inline samples/exitPool/sampleExitBptInForExactTokensOut.json:10]: do you need `minAmount` if it's `BPT_IN_FOR_EXACT_TOKENS_OUT`?

File: samples/exitPool/sampleExitBptInForExactTokensOut.json [hunk 13]
```diff
@@ -1,11 +1,17 @@
...
+    },
+    "0xe0c9275e44ea80ef17579d33c55136b7da269aeb": {
+      "amount": "1",
+      "minAmount": "0.9"
```
Comment (gerrrg) [inline samples/exitPool/sampleExitBptInForExactTokensOut.json:14]: do you need `minAmount` if it's `BPT_IN_FOR_EXACT_TOKENS_OUT`?

File: samples/exitPool/exitPoolSample.py [hunk 14]
```diff
@@ -34,17 +33,16 @@ def main():
         quit()
 
     print()
-    print("===============================================================")
-    print("================= Step 2: Send Bpt to Pool =================")
-    print("===============================================================")
+    print("==============================================================")
+    print("================== Step 2: Send Bpt to Pool ==================")
```
Comment (gerrrg) [inline samples/exitPool/exitPoolSample.py:37]: Didn't catch this on the last review. I get what you mean conceptually, but it's more accurate to say "burn the BPT" though I get that may be confusing for the user.

What do you think about just `Step 2: Exit Pool`?

**Review by gerrrg — CHANGES_REQUESTED** [review]
Looking even better! Just a few small things at this point.

File: balpy/balpy.py (hunk 10 above)
Comment (ribeirojose) [reply balpy/balpy.py:101]: I think this is just the diff being weird 😬

**Comment by ribeirojose** [comment]
@gerrrg ready for another round!

**Pull request merged**